sklearn/manifold/_barnes_hut_tsne.pyx

```diff
@@ -0,0 +1,625 @@
+# cython: boundscheck=False
+# cython: wraparound=False
+# cython: cdivision=True
+# Author: Christopher Moody <chrisemoody@gmail.com>
+# Author: Nick Travers <nickt@squareup.com>
+import numpy as np
+from cython.parallel cimport prange, parallel
+from libc.stdio cimport printf
+cimport numpy as np
+cimport cython
+
+# Implementation by Chris Moody & Nick Travers
+# original code by Laurens van der Maaten
+# for reference implemenations and papers describing the technique:
+# http://homepage.tudelft.nl/19j49/t-SNE.html
+
+
+from libc.stdlib cimport malloc, free, abs
+
+cdef extern from "math.h":
+    double sqrt(double x) nogil
+
+
+# This is effectively an ifdef statement in Cython
+# It allows us to write printf debugging lines
+# and remove them at compile time
+cdef enum:
+    DEBUGFLAG = 0
```

Comment:
An alternative to this would be:

```python

    config.add_extension("_barnes_hut_tsne",
                         sources=["_barnes_hut_tsne.c"],
                         include_dirs=[numpy.get_include()],
                         libraries=libraries,
                         extra_compile_args=["-O3", "-DDEBUGFLAG=0"]) <---
```